src/index/handler/git.rs

```diff
@@ -0,0 +1,349 @@
+use crate::index::handler::{IndexPublishError, IndexPublisher};
+use async_trait::async_trait;
+use forc_pkg::source::reg::{
+    file_location::{location_from_root, Namespace},
+    index_file::{IndexFile, PackageEntry},
+};
+use git2::{Cred, FetchOptions, Oid, PushOptions, RemoteCallbacks, Repository, Signature};
+use std::env;
+use std::fs;
+use std::path::Path;
+use tempfile::TempDir;
+use tokio::task;
+
+/// Index publishing backend for GitHub.
+pub struct GithubIndexPublisher {
+    repo_owner: String,
+    repo_name: String,
+    chunk_size: usize,
+    namespace: Namespace,
+}
+
+impl GithubIndexPublisher {
+    /// Create a new GitHub index publisher.
+    pub fn new(
+        repo_owner: String,
+        repo_name: String,
+        chunk_size: usize,
+        namespace: Namespace,
+    ) -> Self {
+        Self {
+            repo_owner,
+            repo_name,
+            chunk_size,
+            namespace,
+        }
+    }
+
+    fn get_pat_token(&self) -> String {
+        env::var("GITHUB_TOKEN").unwrap()
+    }
+
+    /// Process the GitHub repository and publish the package entry.
+    fn process_repo(&self, package_entry: &PackageEntry) -> Result<(), IndexPublishError> {
+        let repo_url = format!(
+            "https://{}@github.com/{}/{}",
+            self.get_pat_token(),
+            self.repo_owner,
+            self.repo_name
+        );
+
+        // Create a temporary directory
+        let tmp_dir = TempDir::new().map_err(|e| {
+            IndexPublishError::FileSystemError(format!("Failed to create temp dir: {}", e))
+        })?;
+        let tmp_path = tmp_dir.path();
+
+        // Configure callbacks for SSH authentication
+        let callbacks = RemoteCallbacks::new();
+
+        let mut fetch_opts = FetchOptions::new();
+        fetch_opts.remote_callbacks(callbacks);
+
+        let repo = git2::build::RepoBuilder::new()
+            .fetch_options(fetch_opts)
+            .clone(&repo_url, tmp_path)
+            .map_err(|e| {
+                IndexPublishError::CloneError(format!("Failed to clone repository: {}", e))
+            })?;
+
+        let head_ref = repo.head().map_err(|e| {
+            IndexPublishError::UnexpectedError(format!("Failed to get HEAD: {}", e))
+        })?;
+        let branch_name = head_ref.shorthand().unwrap_or("HEAD");
+
+        self.update_and_checkout_default_branch(&repo, branch_name)?;
+        self.write_package_entry(tmp_path, package_entry)?;
+
+        let commit_message = match &self.namespace {
+            Namespace::Flat => format!(
+                "Add package {} version {}",
+                package_entry.name(),
+                package_entry.version()
+            ),
+            Namespace::Domain(domain) => format!(
+                "Add package {}/{} version {}",
+                domain,
```

Comment:
How do domains get handled in the index repo? Does git allow `/` in a directory name?

Reply:
Domains are added as a parent folder. So if there is a domain `SomeDomain` and package is `some`, chunking amount is 2. The resulting file is going to be at `SomeDomain/so/me/some`.